.meta/transforms/lua.toml

```diff
@@ -4,44 +4,147 @@ allow_you_to_description = "transform events with a full embedded [Lua][urls.lua
 beta = true
 common = true
 function_category = "program"
-input_types = ["log"]
-output_types = ["log"]
+input_types = ["log", "metric"]
+output_types = ["log", "metric"]
 requirements = {}
 
-<%= render("_partials/fields/_component_options.toml", type: "transform", name: "lua") %>
-
 [transforms.lua.options.version]
 type = "string"
 common = true
-required = false
-description = "transform API version"
-default = "1"
-enum = { 1 = "transform API version 1" }
+required = true
+description = """
+Transform API version. Indicates the version 2 is used instead of version 1,\
+which is supported, but deprecated.\
+"""
+enum = { 2 = "Lua transform API version 2" }
 
 [transforms.lua.options.source]
 type = "string"
-common = true
+common = false
+required = false
 examples = [
 """\
-require("script") # a `script.lua` file must be in your `search_dirs`
-
-if event["host"] == nil then
-  local f = io.popen ("/bin/hostname")
-  local hostname = f:read("*a") or ""
-  f:close()
-  hostname = string.gsub(hostname, "\\n$", "")
-  event["host"] = hostname
-end\
+custom_module = require('custom_module')\
 """
 ]
-required = true
-description = "The inline Lua source to evaluate."
+description = """
+In version 2 it is the source which is evaluated when the transform is created.\
+In version 1 it is the inline Lua source called for each incoming event.\
+"""
 
 [transforms.lua.options.search_dirs]
 type = "[string]"
 examples = [["/etc/vector/lua"]]
-common = true
+common = false
+required = false
 description = """\
 A list of directories search when loading a Lua file via the `require` \
 function.\
 """
+
+[transforms.lua.options.hooks]
+type = "table"
+category = "Hooks"
+common = true
+required = true
+description = "Configures hooks handlers."
+
+[transforms.lua.options.hooks.children.init]
+type = "string"
+common = false
+required = false
+description = """\
+A function which is called when the first event comes, before calling `hooks.process`\
+"""
+examples = [
+"custom_module.hook_init",
```

Comment:
So if I was a user and wanted to declare these functions (`init()`, `process()`, and `shutdown()`) in some file I write in my favorite LUA editor, then just cite the names here, would it be possible?

I worry this can become a disparity between the wasm work in #2006 as we can't really embed blocks of code in the toml for that. :( I do think that some future `assemblyscript` or other language runtime built on top of wasm would be fine with an implementation like this though.

Reply:
> So if I was a user and wanted to declare these functions (init(), process(), and shutdown()) in some file I write in my favorite LUA editor, then just cite the names here, would it be possible?

Yes, it is possible! It is enough to do the `require` in the `source` section and then just specified the names of imported functions as `custom_module.hook_init`.

> I worry that if we require inlined code here it can become a disparity between the wasm work in #2006 as we can't really embed blocks of code in the toml for that. :( I do think that some future assemblyscript or other language runtime built on top of wasm would be fine with an implementation like this though.

We don't require to inline the code containing the logic, but provide an option to do it. Both loadable modules patterns,  [the one with global functions](https://github.com/timberio/vector/blob/master/rfcs/2020-03-06-1999-api-extensions-for-lua-transform.md#loadable-module-global-functions) and [the one with isolated functions](https://github.com/timberio/vector/blob/master/rfcs/2020-03-06-1999-api-extensions-for-lua-transform.md#loadable-module-isolated-functions) are supported.

Reply:
Great. Thanks Alex. :) Glad we'll have a good UX here.